source-academy/frontend
# Improve Functionality of Remote Execution Device Dialog

### Description

Added a new dependency: `"react-qr-reader": "^3.0.0-beta-1"`

This MR includes various enhancements to the remote execution UI. Mainly, it is done to improve quality-of-life and the out-of-the box experience when pairing with a new device.

Summary of changes:
* Add ability to scan a QR code from the dialog in lieu of manually pasting/typing out the device secret
* Add support for "deep links", where a dialog box (pre-filled) with the device secret will pop up when the accessing SA via the deep link
* Enable the "remote execution" tab in the mobile workspace
* Minor refactoring, code quality improvements and UI fixes

### Type of change


### How to test

* http://localhost:8000/playground should retain existing behaviour, the only difference being on mobile, where the button to show the remote execution tab would also be visible
* http://localhost:8000/playground?add_device=some_string_here_a7Gxa should open the playground directly into the remote execution tab (instead of the home tab), as well as automatically pop-up the add device dialog, pre-filled with the token (`some_string_here_a7Gxa`). At the same time, the address bar will not show the query parameter in the URL.
  * Closing/cancelling/confirming the dialog (or refreshing the page) should not cause it the dialog to pop up again. There was a bug whereby this would happen after resizing a window, when the playground switches between and mobile and desktop workspaces. This was fixed in commit e9d81fb3ca4dc885f980e80f1f3390ae4830095b.
* The add device dialog should now have a button to scan a QR code:
  <img width="554" alt="image" src="https://user-images.githubusercontent.com/34370238/190265760-f1da4e71-49f5-43ba-a830-4530c44f1b3a.png">
  This button should not be visible when editing a device, as the secret token is not editable in the case of editing. After clicking the button and accepting camera permissions, a camera window is shown, and the user can then start to scan the QR code. Clicking the button again will stop the camera. The camera will also auto-close once it detects and scans a QR code.
  <img width="566" alt="image" src="https://user-images.githubusercontent.com/34370238/190266136-b1989916-b8d8-4805-8189-ff8ad4fa539a.png">
  Once the camera has scanned a QR code, it will pre-fill the value into the 'secret' field, overriding any previous value that may be there. This field should remain editable by the user.

### Checklist

- [ ] ~~I have updated the documentation~~ (not applicable)
